README.md

```diff
@@ -1,2 +1,359 @@
+
+```
+
+We can add generating sensors and running the ray-tracing itself to the current flow but
+there is a difference between these new two steps and what we had before. In previous
+example the second step of generating the octree was dependant on the first step and it
+made sense to run them one after another. However we don't need to wait for generating
+sky to generate the sensor grids. The last step of ray-tracing will need both the grid
+and the octree. To describe such flows we will use a Directed Acyclic Graph or DAG. Here
+is the updated flow. Note the the keyword `step` is changed to `tasks` and each `task` has
+a key for `dependency`.
+
+Also since the step for generating grids can generate more than one grid we are using
+loop to run ray-tracing for all these grids in parallel.
+
+```
+
+                              sky    grids
+                                |      |
+                              octree   |
+                                |      |
+                                \______/
+                                    |
+                                ray-tracing
+                         ___________|_____________
+                        /       |       |         \
+                      g-1-1   g-1-2   g-2-1 ... g-n-m
+                        \_______|_______|_________/
+                                    |
+                              collect results
+
+```
+
+```yaml
+
+flow:
+  - tasks:
+    - name: generate_sky_task
+      template: generate_sky
+      inputs:
+        parameters:
+          - name: desired_irradiance
+            value: 100000
+    - name: generate_octree_task
+      template: generate_octree
+      dependencies:
+        - generate_sky_task
+      inputs:
+        parameters:
+           - name: scene_files
+             value:
+               - "{{tasks.generate_sky_steps.outputs.sky_file}}"
+               - "{{workflow.inputs.parameters.folder}}"/model/static/scene.mat
+               - "{{workflow.inputs.parameters.folder}}"/model/static/scene.rad
+    - name: generate_grids_task
+      template: generate_grids
+      inputs:
+        parameters:
+          - name: max_sensor_count
+            value: 200
+    - name: ray_tracing_task
+      template: run_raytracing
+      dependencies:
+        - generate_octree_task
+        - generate_grids_task
+      inputs:
+        parameters:
+          - name: grid
+            value: "{{item}}"
+          - name: octree
+            value: "{{tasks.generate_octree_task.outputs.parameters.octree}}"
+      loop: "{{ tasks.generate_grids_task.outputs.parameters.grids }}"
+```
+
+## 5. outputs
+
+`outputs` is a list of parameters or artifacts that are generated during the flow.
+Several files might be generated in the `flow` section and outputs indicates which ones
+should be considered as the final outputs of the workflow. Executors may return back
+these outputs as a dictionary.
```

Comment:
I think you haven't explained this correctly because I am still confused if `outputs` just refers to the files that are generated during the flow or if `outputs` is a JSON /YAML (or a dictionary?) that notes the output files that you want to save.  I feel like it may be the latter so I would recommend including an example of what this looks like and changing the wording to something like this:

Several files might be generated in the `flow` section and the `outputs` section indicates which ones
should be saved as the final outputs of the workflow. Executors may return these output file paths in a dictionary.

Reply:
I suspect that outputs just like inputs can have a `parameters` and an `artifacts` keys to specify the type of output.

Reply:
Yes and yes, and your wording is much better. I'm going to use it.

I also initially thought of outputs just as files but it can actually be more than just files. For instance a LEED workflow might also return a `parameter` for `passed` which can be `true` or `false`.